runconfig/hostconfig.go

```diff
@@ -17,6 +17,11 @@ func SetDefaultNetModeIfBlank(hc *container.HostConfig) {
 	}
 }
 
+// IsPreDefinedNetwork indicates if a network is predefined by the daemon
+func IsPreDefinedNetwork(network string) bool {
+	return !container.NetworkMode(network).IsUserDefined()
```

Comment:
For a container network, `IsUserDefined() -> false`, so `IsPredefinedNetwork() -> true`.

It was `false` before, which seems right ... I think this needs an extra check?

Reply:
Hm... all the booleans make my head spin. Hm.. so curious why the Windows implementation had it included.

So this function is used on (e.g.) `network create` to prevent one of these networks from being created; that part should still be correct when trying to create a `container:foo` network? So .. wondering if the Linux implementation was faulty; wdyt?

Reply:
Yes, these predicates are fiddly ... I don't think the create/delete functions get called for a container network (there's no separate network), so that's why things are working.

Similarly, I think [filterNetworkByUse](https://github.com/moby/moby/blob/53c521bdc95e7bb5cd59bfac26ea96b1544cad8f/daemon/network/filter.go#L95-L100) and [filterNetworkByType](https://github.com/moby/moby/blob/53c521bdc95e7bb5cd59bfac26ea96b1544cad8f/daemon/network/filter.go#L111-L130) will only be called for the underlying networks. From [Cluster.populateNetworkID](https://github.com/moby/moby/blob/53c521bdc95e7bb5cd59bfac26ea96b1544cad8f/daemon/cluster/networks.go#L316-L317) and [Daemon.localNetworksPrune](https://github.com/moby/moby/blob/53c521bdc95e7bb5cd59bfac26ea96b1544cad8f/daemon/prune.go#L123-L126) there's probably no call for container networks, because the container network doesn't really exist as its own thing. 

So, the change probably wouldn't break anything. But I'd want to stare-at/play-with it some more to be sure, and the behaviour of the function could do with clarifying/unit-testing.

Reply:
Yeah, I think we need to look at cleaning up some of this. Perhaps a single "is reserved name" utility. We currently allow networks on Linux that are reserved names on Windows; I've been thinking about _NOT_ allowing that, and have a single list of reserved ones.

I ran into that case on the CLI side, where some code used "IsUserDefined" on the client-side, but that meant that a Windows CLI connected to a Linux daemon could attempt to create a network with a reserved name. That would of course fail because the API would reject it (so I think I removed that check, and let it be only handled by the API).

Reply:
> I've been thinking about NOT allowing that, and have a single list of reserved ones.

Someone's bound to have a Linux bridge network called "nat". (Maybe someone called Nathan or Natalie!)

Reply:
Yes, that's what kept me from doing so 😂